Implement indirect specular occlusion in StandardMaterial3D

Calinou opened the pull request:
`master` version of https://github.com/godotengine/godot/pull/50603.

Both sources of AO (SSAO and AO texture maps) are used for this effect. The difference is most noticeable when SSAO is enabled (especially with height maps), but this effect also helps when SSAO is disabled.

It is an approximate effect with a very low performance cost, but makes specular reflections more believable overall.

**Testing project:** [test_ao_specular_occlusion_master.zip](https://github.com/godotengine/godot/files/8018158/test_ao_specular_occlusion_master.zip)

## Preview

*SSAO intensity was doubled to make the difference more noticeable. In real world scenarios, the difference will be more subtle.*
*Light Affect is set to 0 (default) both in the SSAO and material settings.*

*Left: partially metallic material (0.5), right: fully metallic material (1.0)*

### SSAO and height enabled

#### Forward+

Before | After
-|-
![Screenshot_20230526_234722 webp](https://github.com/godotengine/godot/assets/180032/e3e88bd1-cd68-4875-8718-62f4187b5c8a) | ![Screenshot_20230526_234703 webp](https://github.com/godotengine/godot/assets/180032/9e89fe48-a042-4d8e-ae22-c3656682e7ff)

#### Mobile[^1]

[^1]: This currently uses a faster, but *supposedly* more approximate approach to indirect specular occlusion.

Before | After
-|-
![Screenshot_20230526_234747 webp](https://github.com/godotengine/godot/assets/180032/f7a586b4-8784-4eff-a477-c381d492bb32) | ![Screenshot_20230526_234735 webp](https://github.com/godotengine/godot/assets/180032/deada2af-646b-43ab-b8dd-a757e78b030f)

### No SSAO and no height (only AO map in material)

#### Forward+

Before | After
-|-
![Screenshot_20230526_234445 webp](https://github.com/godotengine/godot/assets/180032/fc36db5f-8ef9-497a-a35b-8bb0750fc3ef) | ![Screenshot_20230526_234424 webp](https://github.com/godotengine/godot/assets/180032/64fbd2ad-0c0e-4bfa-a683-4dba53b80989)

#### Mobile[^1]

Before | After
-|-
![Screenshot_20230526_234602 webp](https://github.com/godotengine/godot/assets/180032/b103f3fe-a015-47af-95a7-071342e40b8e) | ![Screenshot_20230526_234549 webp](https://github.com/godotengine/godot/assets/180032/0aafec3d-076d-4630-a315-ef09587f5a9d)

[^1]: This currently uses a faster, but *supposedly* more approximate approach to indirect specular occlusion.

Calinou committed:
Implement indirect specular occlusion in StandardMaterial3D

This uses the approximation from Filament. Direct specular lighting
(such as light specular lobes) are not affected.

Both sources of AO (SSAO and AO texture maps) are used for this effect.
The difference is most noticeable when SSAO is enabled (especially with
height maps), but this effect also helps when SSAO is disabled.

It is an approximate effect with a low performance cost, but it makes
specular reflections more believable overall.

clayjohn reviewed (commented):
Looks great as a start!

However, I don't think we should force light_affect to 1.0 for specular lighting. Light_affect is already kind of a specific hack that will only be used by some, I don't think we should force it on all users with no way to turn off. 

An alternative is to separate direct specular from indirect specular and then combine them after you have applied AO to ``specular_light``. Separating the two will also probably be desirable as we polish other rendering features.

Calinou commented:
Superseded by https://github.com/godotengine/godot/pull/51417, which is a simpler implementation that is likely just as effective.

Calinou commented:
Rebased and tested again, it works as expected.

Let me know if I should add a property to adjust specular AO intensity, though I'd prefer the effect to be enabled by default. If the [unused Ao Channel Affect Environment property](https://github.com/godotengine/godot/issues/54611) is removed, this should free up some space in the Environment inspector anyway.

clayjohn commented:
This PR still has the problem that it treats direct and indirect specular the same. It is darkening both direct and indirect by the full amount of ao regardless of the ``ao_light_effect`` setting. AO should only impact indirect light unless a user specifically sets it to impact direct lighting. The Filament docs have a very nice section explaining how specular occlusion can be derived from ambient occlusion: https://google.github.io/filament/Filament.html#lighting/occlusion/specularocclusion

Calinou commented:
> This PR still has the problem that it treats direct and indirect specular the same. It is darkening both direct and indirect by the full amount of ao regardless of the `ao_light_effect` setting. AO should only impact indirect light unless a user specifically sets it to impact direct lighting. The Filament docs have a very nice section explaining how specular occlusion can be derived from ambient occlusion: [google.github.io/filament/Filament.html#lighting/occlusion/specularocclusion](https://google.github.io/filament/Filament.html#lighting/occlusion/specularocclusion)

I managed to separate direct specular lighting from indirect specular lighting and use the Filament approximation. See OP for updated screenshots :slightly_smiling_face: 

However, this appears to have a significant performance cost (0.2 ms of added frame time in my test scene, with SSAO and height mapping disabled).

Note that I still need to port the changes to the mobile renderer.

Calinou commented:
Rebased and tested again, it works as expected. I've updated screenshots in OP. Note that [this comment](https://github.com/godotengine/godot/pull/50601#issuecomment-1031999166) still applies.

Calinou commented:
- Superseded by https://github.com/godotengine/godot/pull/89988.
